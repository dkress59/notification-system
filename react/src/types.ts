import { DetailedHTMLProps, HTMLAttributes, PropsWithChildren } from 'react'

import { SpawnBannerProps, SpawnModalProps, SpawnToastProps } from ':core/types'

type CustomHTMLAttributes<
	T,
	K = void | Record<string, unknown>,
> = DetailedHTMLProps<HTMLAttributes<T> & K, T>

export type BannerAreaAttributes = CustomHTMLAttributes<HTMLBannerAreaElement>

export type NotificationAreaAttributes =
	CustomHTMLAttributes<HTMLNotificationAreaElement>

export type BannerNotificationAttributes =
	CustomHTMLAttributes<HTMLBannerNotificationElement>

export type ModalNotificationAttributes =
	CustomHTMLAttributes<HTMLModalNotificationElement>

export type ToastNotificationAttributes =
	CustomHTMLAttributes<HTMLToastNotificationElement>

declare global {
	// eslint-disable-next-line @typescript-eslint/no-namespace
	namespace JSX {
		interface IntrinsicElements {
			['banner-area']: BannerAreaAttributes
			['notification-area']: NotificationAreaAttributes
			['banner-notification']: BannerNotificationAttributes
			['modal-notification']: ModalNotificationAttributes
			['toast-notification']: ToastNotificationAttributes
		}
	}
}

export type BannerAreaProps = PropsWithChildren

export type NotificationAreaProps = PropsWithChildren

export interface BannerNotificationProps
	extends PropsWithChildren,
		Omit<SpawnBannerProps, 'content'> {
	onDismiss?: () => void
}

export interface ToastNotificationProps
	extends PropsWithChildren,
		Omit<SpawnToastProps, 'content'> {
	onDismiss?: () => void
}

export interface ModalNotificationProps
	extends PropsWithChildren,
		Omit<SpawnModalProps, 'content'> {
	onConfirm?: () => void
	onDecline?: () => void
	onDismiss?: () => void
}
