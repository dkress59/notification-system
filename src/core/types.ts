import {
	HTMLBannerAreaElement,
	HTMLBannerNotificationElement,
	HTMLModalNotificationElement,
	HTMLNotificationAreaElement,
	HTMLToastNotificationElement,
} from './components'

type SnakeToCamelCaseProp<S extends string> = S extends
	| `${infer T} ${infer U}`
	| `${infer T}-${infer U}`
	? `${Uncapitalize<T>}${Capitalize<SnakeToCamelCaseProp<U>>}`
	: Uncapitalize<S>
type SnakeToCamelCase<T> = {
	[K in keyof T as SnakeToCamelCaseProp<K & string>]: T[K]
}

export enum NotificationType {
	SUCCESS = 'success',
	INFO = 'info',
	ERROR = 'error',
	WARNING = 'warning',
}

export enum NotificationEvent {
	BANNER_DISMISSED = 'bannerDismissed',
	MODAL_CONFIRMED = 'modalConfirmed',
	MODAL_DECLINED = 'modalDeclined',
	MODAL_DISMISSED = 'modalDismissed',
	TOAST_DISMISSED = 'toastDismissed',
}

type SpawnArgs<T> = T & {
	content: string
}

export interface BannerAttributes {
	'auto-hide'?: boolean | string
	'auto-hide-after-ms'?: number | string
	'headline'?: string
	'type'?: NotificationType
}
export type SpawnBannerArgs = SpawnArgs<SnakeToCamelCase<BannerAttributes>>

export interface ModalAttributes {
	'condition'?: boolean
	'headline'?: string
	'show-confirm'?: boolean
	'show-decline'?: boolean
	'type'?: NotificationType
	'label-confirm'?: string
	'label-decline'?: string
}
export type SpawnModalArgs = SpawnArgs<SnakeToCamelCase<ModalAttributes>>

export interface ToastAttributes {
	'auto-hide'?: boolean | string
	'auto-hide-after-ms'?: number | string
	'headline'?: string
	'type'?: NotificationType
}
export type SpawnToastArgs = SpawnArgs<SnakeToCamelCase<ToastAttributes>>

declare global {
	interface HTMLElementTagNameMap {
		'banner-area': HTMLBannerAreaElement
		'banner-notification': HTMLBannerNotificationElement
		'modal-notification': HTMLModalNotificationElement
		'notification-area': HTMLNotificationAreaElement
		'toast-notification': HTMLToastNotificationElement
	}
}
