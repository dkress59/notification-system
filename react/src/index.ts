export * from './components/banner-area'
export * from './components/banner-notification'
export * from './components/modal-notification'
export * from './components/notification-area'
export * from './components/toast-notification'
export * from './types'
