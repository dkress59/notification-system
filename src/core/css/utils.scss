%text {
  &-grey {
    color: var(--clr-grey);
  }

  &-green {
    color: var(--clr-green);

    &-light {
      color: var(--clr-green-light);
    }
  }

  &-blue {
    color: var(--clr-blue);

    &-light {
      color: var(--clr-blue-light);
    }
  }

  &-red {
    color: var(--clr-red);

    &-light {
      color: var(--clr-red-light);
    }
  }

  &-yellow {
    color: var(--clr-yellow);

    &-light {
      color: var(--clr-yellow-light);
    }
  }
}

%bg {
  &-grey {
    background-color: var(--clr-grey);
  }

  &-green {
    background-color: var(--clr-green);

    &-light {
      background-color: var(--clr-green-light);
    }
  }

  &-blue {
    background-color: var(--clr-blue);

    &-light {
      background-color: var(--clr-blue-light);
    }
  }

  &-red {
    background-color: var(--clr-red);

    &-light {
      background-color: var(--clr-red-light);
    }
  }

  &-yellow {
    background-color: var(--clr-yellow);

    &-light {
      background-color: var(--clr-yellow-light);
    }
  }
}

%border {
  &-grey {
    border-color: var(--clr-grey);
  }

  &-green {
    border-color: var(--clr-green);

    &-light {
      border-color: var(--clr-green-light);
    }
  }

  &-blue {
    border-color: var(--clr-blue);

    &-light {
      border-color: var(--clr-blue-light);
    }
  }

  &-red {
    border-color: var(--clr-red);

    &-light {
      border-color: var(--clr-red-light);
    }
  }

  &-yellow {
    border-color: var(--clr-yellow);

    &-light {
      border-color: var(--clr-yellow-light);
    }
  }
}

%min-w-60 {
  min-width: 240px;
}

%shadow-lg {
  box-shadow: 0 10px 15px -3px rgb(0 0 0 / 10%), 0 4px 6px -4px rgb(0 0 0 / 10%);
}

%opacity-0 {
  opacity: 0;
}

%opacity-100 {
  opacity: 100;
}
