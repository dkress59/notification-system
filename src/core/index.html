<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0" />
  <title>Notification Web Components</title>

  <script type="module" src="/bundle.js"></script>

  <style>
    * {
      box-sizing: border-box;
    }

    body {
      margin: 0;
      padding: 0;
      font-family: ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
    }
  </style>
  <script>
    document.addEventListener('DOMContentLoaded', () => {
      document.querySelectorAll('toast-notification').forEach(toast => {
        toast.addEventListener('toastDismissed', (event) => console.debug(
          'toast dismissed.',
          { event }
        ))
      })
      document.querySelectorAll('modal-notification').forEach(modal => {
        modal.addEventListener('confirmTriggered', (event) => console.debug(
          'modal confirmed.',
          { event }
        ))
        modal.addEventListener('declineTriggered', (event) => console.debug(
          'modal declined.',
          { event }
        ))
      })

      const bannerArea = document.querySelector('banner-area')
      setTimeout(() => {
        bannerArea.spawnBanner({
          content: 'yoyoyo',
          autoHide: true,
        })
      }, 2000)
      setTimeout(() => {
        bannerArea.spawnBanner({
          content: 'yoyoyo',
          type: 'warning',
          headline: 'spawned',
        })
      }, 4000)

      const notificationArea = document.querySelector('notification-area')
      setTimeout(() => {
        notificationArea.spawnModal({
          content: 'yoyoyo',
          type: 'warning',
          headline: 'spawned',
        })
      }, 1500)
      setTimeout(() => {
        notificationArea.spawnModal({
          content: 'yoyoyo',
          headline: 'spawned new',
        })
      }, 4500)
    })
  </script>
</head>

<body>
  <banner-area>
    <banner-notification headline="some headline" type="error">demo content</toast-notification>
  </banner-area>
  <notification-area>
    <toast-notification class="test" headline="some headline" type="info">demo content</toast-notification>
  </notification-area>
</body>

</html>
