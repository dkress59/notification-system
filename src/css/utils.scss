@tailwind base;
@tailwind components;
@tailwind utilities;

@layer utilities {
  .custom {
    &-toast {
      @apply relative right-0 w-max max-w-1/2 min-w-60 my-4 mx-8;

      &:first-of-type {
        @apply mt-8;
      }
    }

    &-modal {
      @apply fixed top-1/2 left-1/2 w-1/2 h-1/2 min-w-60 max-w-2xl;
      @apply transform-gpu -translate-x-1/2 -translate-y-1/2;
    }

    &-toast,
    &-modal {
      @apply py-4 px-12 pointer-events-auto border-2 rounded-md shadow-lg;
      @apply transition-all duration-500;

      &-type {
        &-success {
          @apply bg-green-light border-green;

          .icon {
            @apply text-green border-green;
          }
        }

        &-info {
          @apply bg-blue-light border-blue;

          .icon {
            @apply text-blue border-blue;
          }
        }

        &-error {
          @apply bg-red-light border-red;

          .icon {
            @apply text-red border-red;
          }
        }

        &-warning {
          @apply bg-yellow-light border-yellow;

          .icon {
            @apply text-yellow border-yellow;
          }
        }
      }

      >.icon {
        @apply absolute top-4 left-4 h-6 aspect-square rounded-full overflow-hidden border-2;
        @apply text-center text-xs leading-5 font-bold;
      }

      >h4 {
        @apply font-bold text-lg text-grey leading-snug;
      }

      >div {
        @apply text-grey;
      }

      >.dismiss {
        @apply absolute top-4 right-4 font-bold cursor-pointer;
      }
    }
  }
}
