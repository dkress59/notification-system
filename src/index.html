<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0" />
  <title>Notification Web Components</title>

  <script type="module" src="/build/notification-system.esm.js"></script>
  <script nomodule src="/build/notification-system.js"></script>

  <style>
    * {
      box-sizing: border-box;
    }

    body {
      font-family: sans-serif;
    }

    #notification-container {
      position: fixed;
      z-index: 50;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      pointer-events: none;
      display: flex;
      flex-direction: column;
      align-items: flex-end;
      overflow-y: auto;
    }

    /* #notification-container > * {
        right: 0;
      } */
  </style>
  <script>
    document.addEventListener('DOMContentLoaded', () => {
      document.querySelectorAll('notify-toast').forEach(toast => {
        toast.addEventListener('toastDismissed', (shadowRoot) => console.debug(
          'toast dismissed!',
          { shadowRoot }
        ))
      })
    })
  </script>
</head>

<body>
  <div id="notification-container">
    <!-- <my-component first="Stencil" last="'Don't call me a framework' JS"></my-component>
      <my-component first="more Stencil" middle="please" last="'Don't call me a framework' JS"></my-component> -->
    <notify-toast type="error" headline="Lorem Ipsum">Lorem ipsum dolor sit amet consectetur adipisicing elit, facilis
      necessitatibus sunt quisquam officia quidem.</notify-toast>
    <notify-toast>Mock Content</notify-toast>
    <notify-toast auto-hide>Mock Content</notify-toast>
    <notify-toast type="info" is-hidden>Hidden Content</notify-toast>
    <notify-toast>Mock Content</notify-toast>
  </div>
</body>

</html>