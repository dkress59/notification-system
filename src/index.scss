@import 'css/base.scss';
// @import 'css/util.scss';
