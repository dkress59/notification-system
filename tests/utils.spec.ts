import { newSpecPage } from '@stencil/core/testing'

import { NotificationType } from '../src/types'
import { getButton, getHeadline, getIcon } from '../src/utils'

describe('utils', () => {
	describe('getButton', () => {
		it('matches snapshot', async () => {
			const page = await newSpecPage({
				components: [],
				template: () => getButton(jest.fn()),
			})
			expect(page.root).toMatchSnapshot()
		})
		it('fires custom onClick event', async () => {
			const mockEventHandler = jest.fn()
			const page = await newSpecPage({
				components: [],
				template: () => getButton(mockEventHandler),
			})
			const button = page.root!
			expect(button).toBeTruthy()
			button.dispatchEvent(new Event('click'))
			expect(mockEventHandler).toHaveBeenCalled()
		})
	})
	describe('getHeadline', () => {
		it('returns undefined if no headline is provided', () => {
			expect(getHeadline()).toBeUndefined()
		})
		it('matches snapshot if a headline is provided', async () => {
			const page = await newSpecPage({
				components: [],
				template: () => getHeadline('Mock Headline'),
			})
			expect(page.root).toMatchSnapshot()
		})
	})
	describe('getIcon', () => {
		it('type success matches snapshot', async () => {
			const page = await newSpecPage({
				components: [],
				template: () => getIcon(NotificationType.SUCCESS),
			})
			expect(page.root).toMatchSnapshot()
		})
		it('type error matches snapshot', async () => {
			const page = await newSpecPage({
				components: [],
				template: () => getIcon(NotificationType.ERROR),
			})
			expect(page.root).toMatchSnapshot()
		})
		it('type info matches snapshot', async () => {
			const page = await newSpecPage({
				components: [],
				template: () => getIcon(NotificationType.INFO),
			})
			expect(page.root).toMatchSnapshot()
		})
		it('type warning matches snapshot', async () => {
			const page = await newSpecPage({
				components: [],
				template: () => getIcon(NotificationType.WARNING),
			})
			expect(page.root).toMatchSnapshot()
		})
	})
})
