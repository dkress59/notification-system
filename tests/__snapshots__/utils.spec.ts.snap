// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils getButton matches snapshot 1`] = `
<span class="dismiss" data-testid="dismiss-button">
  ✕
</span>
`;

exports[`utils getHeadline matches snapshot if a headline is provided 1`] = `
<h4 data-testid="headline">
  Mock Headline
</h4>
`;

exports[`utils getIcon type error matches snapshot 1`] = `
<span class="icon" data-testid="notify-icon">
  <span class="error">
    ╲
  </span>
</span>
`;

exports[`utils getIcon type info matches snapshot 1`] = `
<span class="icon" data-testid="notify-icon">
  i
</span>
`;

exports[`utils getIcon type success matches snapshot 1`] = `
<span class="icon" data-testid="notify-icon">
  <span>
    ✓
  </span>
</span>
`;

exports[`utils getIcon type warning matches snapshot 1`] = `
<span class="icon" data-testid="notify-icon">
  !
</span>
`;
